apps/meteor/client/lib/chats/readStateManager.ts

```diff
@@ -0,0 +1,174 @@
+		callback(this.firstUnreadRecordId);
+		const unsub = this.on('unread-state-change', callback);
+		return unsub;
+	};
+
+	public updateSubscription(subscription?: ISubscription) {
+		if (!subscription) {
+			return;
+		}
+
+		const firstUpdate = !this.subscription;
+
+		this.subscription = subscription;
+		LegacyRoomManager.getOpenedRoomByRid(this.rid)?.unreadSince.set(this.subscription.ls);
+
+		const { unread, alert } = this.subscription;
+		if (!unread && !alert) {
+			return;
+		}
+
+		if (firstUpdate) {
+			this.updateFirstUnreadRecordId();
+			return;
+		}
+
+		if (document.hasFocus() && this.firstUnreadRecordId) {
+			return;
+		}
+
+		this.updateFirstUnreadRecordId();
+	}
+
+	private updateFirstUnreadRecordId() {
+		if (!this.subscription?.ls) {
+			return;
+		}
+
+		const firstUnreadRecord = ChatMessage.findOne(
+			{
+				'rid': this.subscription.rid,
+				'ts': {
+					$gt: this.subscription.ls,
+				},
+				'u._id': {
+					$ne: Meteor.userId() ?? undefined,
+				},
+			},
+			{
+				sort: {
+					ts: 1,
+				},
+			},
+		);
+
+		this.setFirstUnreadRecordId(firstUnreadRecord?._id);
+
+		RoomHistoryManager.once('loaded-messages', () => this.updateFirstUnreadRecordId());
+	}
+
+	private setFirstUnreadRecordId(firstUnreadRecordId: string | undefined) {
+		this.firstUnreadRecordId = firstUnreadRecordId;
+		this.emit('unread-state-change', this.firstUnreadRecordId);
+	}
+
+	public clearUnreadMark() {
+		this.setFirstUnreadRecordId(undefined);
+	}
+
+	public handleWindowEvents = (): UnsuscribeWindowEvents => {
+		const handleWindowFocus = () => {
+			this.attemptMarkAsRead();
+		};
+
+		const handleWindowKeyUp = (event: KeyboardEvent) => {
+			if (event.key === 'Escape') {
+				this.markAsRead();
+				this.updateFirstUnreadRecordId();
+			}
+		};
+
+		window.addEventListener('focus', handleWindowFocus);
+		window.addEventListener('keyup', handleWindowKeyUp);
+
+		return () => {
+			window.removeEventListener('focus', handleWindowFocus);
+			window.removeEventListener('keyup', handleWindowKeyUp);
+		};
+	};
+
+	private isUnreadMarkVisible(): boolean {
+		if (!this.unreadMark) {
+			return false;
+		}
+
+		return this.unreadMark.offsetTop > (this.unreadMark.offsetParent?.scrollTop || 0);
+	}
+
+	// This will only mark as read if the unread mark is visible
+	public attemptMarkAsRead() {
+		const { alert, unread } = this.subscription || {};
+		if (!alert && unread === 0) {
+			return;
+		}
+
+		if (!document.hasFocus()) {
+			return;
+		}
+
+		if (this.unreadMark && !this.isUnreadMarkVisible()) {
+			return;
+		}
+		// if there are unloaded unread messages, don't mark as read
+		if (RoomHistoryManager.getRoom(this.rid).unreadNotLoaded.get() > 0) {
+			return;
+		}
+
+		return this.markAsRead();
+	}
+
+	public debouncedMarkAsRead = withDebouncing({ wait: 1000 })(() => {
+		try {
+			return this.markAsRead();
+		} catch (e) {
+			console.error(e);
```

Comment:
`question` Why catch this error just to log it?

Reply:
The `subscriptions.read` api call is very strict about sending too many requests. I added a debounce to it, but it can still trigger the delay. If `subscriptions.read` endpoint is blocked and I send a message, it blocks the message from being sent until the endpoint is active again (takes about 30s or longer). This way, if it happens to trigger the rate limiter, I can still send messages without issues.

Also, this is very similar to the current behaviour, since a bunch of too-many-requests errors pop in the console, but I can send messages normally.